torchdata/datapipes/iter/util/webdataset.py

```diff
@@ -0,0 +1,69 @@
+import re
+from typing import Dict, Iterator, List, Union
+
+from torchdata.datapipes import functional_datapipe
+from torchdata.datapipes.iter import IterDataPipe
+
+
+def pathsplit(p):
+    """Split a path into the basename and the extensions."""
+    if "." not in p:
+        return p, ""
+    prefix, suffix = re.search(r"^(.*?)(\.[^/]*)$", p).groups()
+    return prefix, suffix
+
+
+@functional_datapipe("webdataset")
+class WebDatasetIterDataPipe(IterDataPipe[Dict]):
+    r"""
+    Iterable DataPipe that accepts stream of (path, data) tuples (usually,
+    representing the pathnames and files of a tar archive) and aggregates
+    consecutive items with the same basename into a single dictionary, using the
+    extensions as keys (WebDataset file convention). Any text after the first
+    "." in the filename is used as an a key/extension.
+
+    File names that do not have an extension are ignored.
+
+    Args:
+        source_datapipe: a DataPipe yielding a stream of (path, data) pairs
+    Returns:
+        a DataPipe yielding a stream of dictionaries
+    Examples:
+        >>> from torchdata.datapipes.iter import FileLister, FileOpener
+        >>>
+        >>> def decode(item):
+        >>>     key, value = item
+        >>>     if key.endswith(".txt"):
+        >>>         return key, value.read().decode("utf-8")
+        >>>     if key.endswith(".bin"):
+        >>>         return key, value.read().decode("utf-8")
+        >>>
+        >>> datapipe1 = FileLister("test/_fakedata", "wds*.tar")
+        >>> datapipe2 = FileOpener(datapipe1, mode="b")
+        >>> dataset = datapipe2.load_from_tar().map(decode).webdataset()
+        >>> for obj in dataset:
+        >>>     print(obj)
+    """
+
+    def __init__(self, source_datapipe: IterDataPipe[List[Union[Dict, List]]]) -> None:
+        self.source_datapipe: IterDataPipe[List[Union[Dict, List]]] = source_datapipe
+
+    def __iter__(self) -> Iterator[Dict]:
+        sample = {}
+        current = ""
+        for path, data in self.source_datapipe:
+            assert isinstance(path, str), path
+            prefix, suffix = pathsplit(path)
+            if suffix == "":
+                continue
```

Comment:
nit: How often will people run into this? And if they do, is it because `pathsplit` made a mistake or the users named the files incorrectly? I am wondering if we should raise an exception or warning instead.

Reply:
Files without extensions can be used for metadata; it's fine for readers to ignore them. 

Since users need an extension to refer to a file anyway, there really is no risk of having file names without extensions accidentally; if they generated such files, they would immediately realize that there is no way to refer to the file inside the sample.

Reply:
Makes sense to me. Do you mind adding a short inline comment there for future reference?

Reply:
Done.